linera-service/src/project.rs

```diff
@@ -192,7 +193,65 @@ impl Project {
         debug!("creating cargo config");
         Self::create_cargo_config(&root)?;
 
-        Ok(Self { _root: root })
+        Ok(Self { root })
+    }
+
+    pub fn from_existing_project(root: PathBuf) -> Result<Self> {
+        if !root.exists() {
+            bail!("could not find project at {}", root.display());
+        }
+        Ok(Self { root })
+    }
+
+    pub fn test(&self) -> Result<()> {
+        if !Self::runner_is_installed()? {
+            debug!("Linera test runner not found");
+            Self::install_test_runner()?;
+        }
+        let cargo_test = Command::new("cargo")
+            .env(
+                "CARGO_TARGET_WASM32_UNKNOWN_UNKNOWN_RUNNER",
+                Self::runner_path()?.display().to_string().as_str(),
+            )
+            .arg("test")
+            .args(["--target", "wasm32-unknown-unknown"])
+            .current_dir(&self.root)
+            .spawn()?
+            .wait()?;
+        if !cargo_test.success() {
+            bail!("tests failed")
+        }
+        Ok(())
+    }
+
+    pub fn runner_is_installed() -> Result<bool> {
+        Ok(Self::runner_path()?.exists())
+    }
+
+    fn install_test_runner() -> Result<()> {
+        debug!("installing test runner");
```

Comment:
Or maybe `println!`? I'd want to know when something gets automatically installed.

(Personally, I'd even prefer if it didn't, and just told the user to install it. But it's more convenient this way, of course.)